collector/collector.go

```diff
@@ -50,28 +54,100 @@ type ReportTemplate struct {
 	TestVersion string `json:"test_version"`
 }
 
+// PrivacySettings contains privacy settings for submitting measurements.
+type PrivacySettings struct {
+	// IncludeASN indicates whether to include the ASN
+	IncludeASN bool
+
+	// IncludeCountry indicates whether to include the country
+	IncludeCountry bool
+
+	// IncludeIP indicates whether to include the IP
+	IncludeIP bool
+}
+
+func (ps PrivacySettings) apply(m *model.Measurement) (err error) {
+	if ps.IncludeASN == false {
+		m.ProbeASN = model.DefaultProbeASNString
+	}
+	if ps.IncludeCountry == false {
+		m.ProbeCC = model.DefaultProbeCC
+	}
+	if ps.IncludeIP == false {
+		err = ps.maybeRewriteTestKeys(m) // MUST be before the following statement
+		m.ProbeIP = model.DefaultProbeIP
+	}
+	return
+}
+
+func (ps PrivacySettings) maybeRewriteTestKeys(m *model.Measurement) error {
+	if net.ParseIP(m.ProbeIP) == nil {
+		return errors.New("Invalid m.ProbeIP string")
+	}
+	data, err := json.Marshal(m.TestKeys)
+	if err != nil {
+		return err
+	}
+	// The check using Count is to save an unnecessary copy performed by
+	// ReplaceAll when there are no matches into the body. This is what
+	// we would like the common case to be, meaning that the code has done
+	// its job correctly and has not leaked the IP.
+	pip := []byte(m.ProbeIP)
```

Comment:
I am not very convinced by this pattern here as it's dependent on the order in which functions are getting called and by inverting the order you run the risk of not scrubbing the right address.

I would be more confident if instead we passed to the `PrivacySettings` structure the `LocationInfo` (https://github.com/ooni/probe-engine/blob/master/model/model.go#L104) struct which is something that we do not need to redact or strip.

This way the scrubbing is done independently from the values inside of the object you are scrubbing.

Reply:
That would break resubmission of a report saved on disk, where `LocationInfo` would be the relative to the resubmitting probe rather than to the measuring probe.

If the concern is that of ordering, which is a wise concern, I can refactor the code such that the concern is completely addressed. Will push a diff in a second.

Reply:
That a good point, I had not considered. I do wonder then if it's wise to do the scrubbing in the moment of submission of the report and rather not do it when you create the measurement itself and write it to disk.

Doing it in the moment of collection also raises the question of how to interpret the privacy settings. I believe the privacy settings of collecting or not the probe IP should be relevant to when the user triggers the test and not when they are uploading them (and I believe this is also what we have been doing historically).

As it is now if the user is in a risky situation and they have disabled to collect their IP, then they re-enable it when they get back home they probably still don't want the measurements from the risky situation to leak their IP even if they later changed the settings in a different context.

Reply:
So, I think I have found a better pattern that should be robust with respect to the concern you have expressed, and additionally should not break with resubmission.

Reply:
> . I do wonder then if it's wise to do the scrubbing in the moment of submission of the report [...]

Yes, that's the right thing to do!